### DISCUSSION TITLE
JsonRPC logging method

### DISCUSSION THREAD

**Discussion by Garulf** [Flow-Launcher/Flow.Launcher#632] [Ideas]
Is it possible to implement a method for logging directly to flow's log file?

Currently all stdout output invokes a Json parsing error. stderr output writes to the log but halts the plugin.

At this time [`debug`](https://github.com/Flow-Launcher/Flow.Launcher.JsonRPC.Python/blob/1e95f0df1853bee026c3bfea64ff48ea133db884/flowlauncher/FlowLauncher.py#L47)  method returns a parsing error as well.

Best case scenario all output that's not Json just gets added to the log

**Comment by taooceros** [comment]
You can take a try on using logging, but we haven't tested it for now.

To do so, pass a callback method `Flow.Launcher.LogInfo` with a object[] containing three parameter `string className, string message, [CallerMemberName] string methodName = ""`.

**Comment by taooceros** [comment]
We also have support on debug message, but that's not current available in the python package written by @Zeroto521. To use that, you may need to manually `print` the json parsed message with a property `debugmessage` (we read your output via stdout).

```json
{
"result": "list of result",
"debugMessage": "message"
}
```

**Comment by taooceros** [comment]
By the way, do you think we shouldn't halt the plugin if stdout & stderr both contains message?

**Comment by Garulf** [comment]
> By the way, do you think we shouldn't halt the plugin if stdout & stderr both contains message? 

I was thinking about it. On one hand it would be nice not to halt so the plugin dev could list a result to handle the error (like an option to try again or open a GitHub issue). But if it's an unhandled error and the stdout is garbage how does the launcher handle that?

**Reply by taooceros** [reply]
Currently it will only log it into log. JsonRPC error handling is a bit different from other plugin, so it won't pop out an error window for now (will modify it in the future).

I think unhandled error will mean no stdout, which will be different from intended stderr.

Here's an example of invoking debug message.

```python
# -*- coding: utf-8 -*-

import json


output = {
    "result": [{
        "title": "test",
        "subtitle": ""
    }],
    "debugMessage": "test"
}

print(json.dumps(output, indent=False))
```

**Reply by Garulf** [reply]
> Currently it will only log it into log. JsonRPC error handling is a bit different from other plugin, so it won't pop out an error window for now (will modify it in the future).
> 
> 
> 
> I think unhandled error will mean no stdout, which will be different from intended stderr.
> 
> 
> 
> Here's an example of invoking debug message.
> 
> 
> 
> ```python
> 
> # -*- coding: utf-8 -*-
> 
> 
> 
> import json
> 
> 
> 
> 
> 
> output = {
> 
>     "result": [{
> 
>         "title": "test",
> 
>         "subtitle": ""
> 
>     }],
> 
>     "debugMessage": "test"
> 
> }
> 
> 
> 
> print(json.dumps(output, indent=False))
> 
> ```

Okay, makes sense. 

Having proper debug messages will help a ton. Thank you.

**Comment by Zeroto521** [comment]
In my experience, there have two debug methods at present.

1. Just run the plugin entry script in commander (ex `python main.py`) with the `print` function. I thought this is the easiest way.
2. Don't forget to use python's `logging` package or `open('plugin.log', 'w')` function. You could totally do your own plugin log with this method. I don't suggest gather plugin log information with flow log information. It is better to sperate plugin its and flow its log, don't mix them up.

**Reply by jjw24** [reply]
>  I don't suggest gather plugin log information with flow log information. It is better to sperate plugin its and flow its log, don't mix them up.

It might not be viable as @Garulf mentioned below as well, splitting logging up creates confusion for users to retrieve. Any logging including from plugins that enables users to troubleshoot or requires user upload so we can debug, should go into one log, anything else for development logging, you can do your own logging else where.

**Comment by Garulf** [comment]
> In my experience, there have two debug methods at present.
> 
> 
> 
> 1. Just run the plugin entry script in commander (ex `python main.py`) with the `print` function. I thought this is the easiest way.
> 

With this method you could even provide a `-v` flag for your plugin. Good for debugging but not great for the end user. 

> 2. Don't forget to use python's `logging` package or `open('plugin.log', 'w')` function. You could totally do your own plugin log with this method. I don't suggest gather plugin log information with flow log information. It is better to sperate plugin its and flow its log, don't mix them up.

I think some log output as long as it's labeled properly will be a good thing. Any user having an issue need only look at one file. 
I think of Home Assistant's logging method for example. All components inherit one logger.
